modules/catalog-compaction/src/main/java/org/apache/ignite/internal/catalog/compaction/CatalogCompactionRunner.java

```diff
@@ -387,24 +435,61 @@ private CompletableFuture<Boolean> tryCompactCatalog(Catalog catalog, LogicalTop
                     }
 
                     return catalogManagerFacade.compactCatalog(catalog.version());
+                }).whenComplete((res, ex) -> {
+                    if (ex != null) {
+                        LOG.warn("Catalog compaction has failed [timestamp={}].", ex, minRequiredTime);
+                    } else {
+                        if (res) {
+                            LOG.info("Catalog compaction completed successfully [timestamp={}].", minRequiredTime);
+                        } else {
+                            LOG.info("Catalog compaction skipped [timestamp={}].", minRequiredTime);
+                        }
+                    }
                 });
     }
 
-    private CompletableFuture<Set<String>> requiredNodes(Catalog catalog) {
+    private CompletableFuture<Pair<Boolean, Set<String>>> validatePartitions(
+            Catalog catalog,
+            Map<Integer, BitSet> localPartitions,
+            Map<String, Map<Integer, BitSet>> remotePartitions
+    ) {
         HybridTimestamp nowTs = clockService.now();
-        Set<String> required = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        ConcurrentHashMap<String, CurrentPartitions> currentPartitionsPerNode = new ConcurrentHashMap<>();
 
         return CompletableFutures.allOf(catalog.tables().stream()
-                .map(table -> collectRequiredNodes(catalog, table, required, nowTs))
+                .map(table -> collectRequiredNodes(catalog, table, nowTs, currentPartitionsPerNode))
                 .collect(Collectors.toList())
-        ).thenApply(ignore -> required);
+        ).thenApply(ignore -> {
+
+            Map<String, Map<Integer, BitSet>> all = new HashMap<>(remotePartitions);
+            all.put(localNodeName, localPartitions);
+
+            Set<String> required = currentPartitionsPerNode.keySet();
+
+            for (Map.Entry<String, Map<Integer, BitSet>> node : all.entrySet()) {
+                String nodeId = node.getKey();
+                Map<Integer, BitSet> tables = node.getValue();
+                CurrentPartitions partitionsPerNode = currentPartitionsPerNode.get(nodeId);
```

Comment:
looks like case when `currentPartitionsPerNode` has more nodes than we collected is missing (we need test for this case)

Reply:
Fixed.